pkg/local_object_storage/metabase/graveyard.go

```diff
@@ -259,3 +260,104 @@ func (db *DB) DropGraves(tss []TombstonedObject) error {
 		return nil
 	})
 }
+
+// GetGarbage returns garbage according to the metabase state. Garbage includes
+// objects marked with GC mark (expired, tombstoned but not deleted from disk,
+// extra replicated, etc.) and removed containers.
+// The first return value describes garbage objects. These objects should be
+// removed. The second return value describes garbage containers whose _all_
+// garbage objects were included in the first return value and, therefore,
+// these containers can be deleted (if their objects are handled and deleted too).
+func (db *DB) GetGarbage(limit int) ([]oid.Address, []cid.ID, error) {
+	if limit <= 0 {
+		return nil, nil, nil
+	}
+
+	db.modeMtx.RLock()
+	defer db.modeMtx.RUnlock()
+
+	if db.mode.NoMetabase() {
+		return nil, nil, ErrDegradedMode
+	}
+
+	const reasonableLimit = 1000
+	initLimit := reasonableLimit
+	if limit < initLimit {
+		initLimit = limit
+	}
+
+	var addrBuff oid.Address
+	var cidBuff cid.ID
+	var uniqueObjectsMap map[oid.ID]struct{}
+	alreadyHandledContainers := make(map[cid.ID]struct{})
+	resObjects := make([]oid.Address, 0, initLimit)
+	resContainers := make([]cid.ID, 0)
+
+	err := db.boltDB.View(func(tx *bbolt.Tx) error {
+		// start from the deleted containers since
+		// an object can be deleted manually but
+		// also be deleted as a part of non-existing
+		// container so no need to handle it twice
+
+		bkt := tx.Bucket(garbageContainersBucketName)
+		c := bkt.Cursor()
+
+		for k, _ := c.First(); k != nil; k, _ = c.Next() {
+			err := cidBuff.Decode(k)
+			if err != nil {
+				return fmt.Errorf("parsing raw CID: %w", err)
+			}
+
+			// another container, clean the map
+			uniqueObjectsMap = make(map[oid.ID]struct{}, initLimit)
+
+			err = listContainerObjects(tx, cidBuff, uniqueObjectsMap, limit-len(resObjects))
+			if err != nil {
+				return fmt.Errorf("listing objects for %s container: %w", cidBuff, err)
+			}
+
+			addrBuff.SetContainer(cidBuff)
+			for obj := range uniqueObjectsMap {
+				addrBuff.SetObject(obj)
+				resObjects = append(resObjects, addrBuff)
+			}
+
+			alreadyHandledContainers[cidBuff] = struct{}{}
+
+			if len(resObjects) < limit {
```

Comment:
you mean it would save us if `listContainerObjects` is buggy? it is an unexported func with its own `limit` arg, i think we should not duplicate checks because of that

Reply:
no, i missed that `listContainerObjects` limit is updated with `resObjects ` progress, so my comment is irrelevant

besides this: can't we add `cidBuff` to the result if `len(resObjects) == limit`?

Reply:
in the current design -- no. we cant say for sure if it is a limit restriction (and we have more objects but we wont get them for this iteration) or we just dont have any more objects. can be fixed a little with some bool flags and more complex code but i do not think it is worth it: that should be a really rare case with no bad affects (container will be deleted just in the next iteration)